### PR TITLE
More ntp servers

### PR DESCRIPTION
**Pull request by christoph-zededa**
- allow setting more than one NTP server
- allow setting NTP servers by domain (instead of just ip)
- allow ignoring NTP servers from DHCP server

### CODE REVIEW

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (milan-zededa) [inline pkg/pillar/cmd/zedrouter/networkinstance.go:124]: zedrouter has its own logger, use `z.log`

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (milan-zededa) [inline pkg/pillar/cmd/zedrouter/networkinstance.go:119]: I think that you can just pass `devicenetwork.ResolveCacheWrap` as the argument, no need to wrap it.

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedrouter/networkinstance.go]: I don't really like boolean flags, so I did this instead: https://github.com/lf-edge/eve/pull/4448/commits/5f34b338fba8299308c0065f1cd7625df5e7cea7

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (milan-zededa) [inline pkg/pillar/cmd/zedrouter/networkinstance.go:140]: I think it would be less confusing if getNIPortConfig didn't set NTPServers at all (not even the already resolved IPs) and would leave this to attachNTPServersToPortConfigs.
Then also please add a comment for getNIPortConfig mentioning that NTP servers are set separately.

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (milan-zededa) [reply pkg/pillar/cmd/zedrouter/networkinstance.go]: ok

**Comment by OhmSpectator** [comment]
Argh, I will try to review it soon...

File: Makefile
Comment (rene) [inline Makefile:252]: you can put this definition under a `else` in the if below and avoid set this variable (unnecessarily) when TAP is not null...

File: pkg/pillar/cmd/zedagent/parseconfig.go
Comment (rene) [inline pkg/pillar/cmd/zedagent/parseconfig.go:2308]: `config.NtpServers` will never be null here because it's being initialized right above when it's null....

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedrouter/networkinstance.go]: done

File: Makefile
Comment (christoph-zededa) [reply Makefile]: done

File: pkg/pillar/cmd/zedagent/parseconfig.go
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go]: done

File: pkg/pillar/cmd/zedagent/reportinfo.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedagent/reportinfo.go:1045]: You seem to be overwriting this devicePort.NtpServer on line 1053.  Isn't the intent to collect all the ntp servers in the ntpServers variable and then update the devicePort below?

File: pkg/pillar/types/dns.go
Comment (eriknordmark) [inline pkg/pillar/types/dns.go:589]: Is this FilterDuplicates across a mixture of IP address strings and domain names (which might resolve to some IP addresses already in the list) required? helpful? confusing?

The key filter is down in getIntendedDnsmasqCfg where we make filter out the duplicates based on the resolved IP addresses.

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedrouter/networkinstance.go:108]: The timeout is specified here: https://github.com/lf-edge/eve/blob/master/pkg/pillar/devicenetwork/dns.go#L35

Last week I tested the code with more than 10 unreachable NTP servers and I did not notice the watchdog to get triggered.

Do we run `zedrouter` without watchdog? - https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/zedrouter/zedrouter.go#L175 as I don't see a `agentbase.WithWatchdog(...)`.

File: pkg/pillar/cmd/zedagent/reportinfo.go
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedagent/reportinfo.go]: Yes, this code is wrong. It should be fixed now.

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go:111]: In the end `chronyd` and `dnsmasq` do not care about the different interfaces, do they?

Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go]: I made the map key a combination of domain and source ip.

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/networkinstance.go:108]: We kick of the main watchdog for each agent in device-steps.sh
    touch "$WATCHDOG_FILE/$AGENT.touch"

Wouldn't the delay be if there are potentially unreachable/slow DNS servers and there are 10 NTP servers configured with a domain name that needs to be resolved?

File: pkg/pillar/devicenetwork/dns.go
Comment (eriknordmark) [reply pkg/pillar/devicenetwork/dns.go]: dnsmasq is for the applications doing DNS resolution. We configure chronyd with IP addresses. My concern is about the DNS resolution of the  NTP server names. Note that we have been resolving the standard pool.ntp.org without worrying about walled gardens, but as we allow the users to specify domain names for NTP servers they might be using ones which are inside their walled garden(s). I'll take a look at your map approach.

**Review by eriknordmark — APPROVED** [review]
LGTM but still need to see if we the DNS resolutions of lots of names can take too long for the watchdog.
Kick off tests.

File: pkg/pillar/cmd/zedrouter/networkinstance.go
Comment (christoph-zededa) [reply pkg/pillar/cmd/zedrouter/networkinstance.go:109]: I think, now I understand (better) how this works.
I added a call to `pubSub.StillRunning(...)`: https://github.com/lf-edge/eve/compare/2a16d63887c83fbe4ffe5bec74b3a435b0f7f5f2..b9a29770fc4d1103a007ba1bbe3286172047bf4f

File: pkg/pillar/devicenetwork/dns.go
Comment (christoph-zededa) [reply pkg/pillar/devicenetwork/dns.go]: I also added a test for the map approach (perhaps that's easier to understand): https://github.com/lf-edge/eve/pull/4448/files#diff-6ecf3e43642f91236a8006c29a4a1abfe7d1ef210b6381ff39808acd3c64bd90R280

File: pkg/pillar/types/dns.go
Comment (christoph-zededa) [reply pkg/pillar/types/dns.go:591]: Sure, I did.

**Comment by christoph-zededa** [comment]
> Kick off tests.

Sorry, I saw that too late and force-pushed ...

**Pull request merged**
